### PR TITLE
Edges and Overlays Optimizations

### PR DESCRIPTION
**Pull request by LeandroTreu**
- Sorting connectors so that edges don't intertwine
- Summarize edges with a symbol in cases where too many are rendered
- Hovering over nodes shows all attached edges
- Adjusted nodesep and ranksep
- Fixed error when selecting memlet
- Added Shift+Rightclick to expand/collapse
- Added Ctrl+Scroll verticalscroll in zoom mode
- Various Overlay LOD threshold adjustments

## Before
edges_unsorted

## Sorted
edges_sorted

## Summarized
edges_summarised

### CODE REVIEW

**Comment by tbennun** [comment]
I think that the overlay LOD management needs to be refactored out to some node overlay superclass.

File: src/renderer/renderer_elements.ts [hunk 1]
```diff
@@ -54,6 +54,19 @@ export class SDFGElement {
     public selected: boolean = false;
     public highlighted: boolean = false;
     public hovered: boolean = false;
+    
+    // Used to draw edge summary instead of all edges separately.
+    // Helps with rendering performance when too many edges would be drawn on the screen.
+    // These two fields get set in the layouter, depending on the number of in/out_connectors
+    // of a node. They also get toggled in the mousehandler when the hover status changes.
+    // Currently only used for NestedSDFGs and ScopeNodes.
+    public summarise_in_edges: boolean = false;
```
Comment (tbennun) [inline src/renderer/renderer_elements.ts:63]: In the dace codebase, we use American English for spelling.

File: src/sdfv.ts [hunk 2]
```diff
@@ -528,7 +530,7 @@ function file_read_complete(sdfv: SDFV): void {
             sdfv.get_renderer()?.destroy();
             sdfv.set_renderer(new SDFGRenderer(sdfv, sdfg, container, mouse_event));
             sdfv.close_menu();
-        }, 20);
+        }, 10);
```
Comment (tbennun) [inline src/sdfv.ts:533]: why this change?

File: src/renderer/renderer.ts [hunk 3]
```diff
@@ -856,9 +857,35 @@ export class SDFGRenderer extends EventEmitter {
...
+        // Initial layout fully collapsed
+        this.for_all_sdfg_elements(
+            (_t: SDFGElementGroup, _d: any, obj: any) => {
+                if ('is_collapsed' in obj.attributes &&
+                    !obj.type.endsWith('Exit'))
+                    obj.attributes.is_collapsed = true;
+            }
+        );
+        this.emit('collapse_state_changed', true, true);
+        this.relayout();
```
Comment (tbennun) [inline src/renderer/renderer.ts:874]: Why relayout twice?

File: src/renderer/renderer.ts [hunk 4]
```diff
@@ -856,9 +857,35 @@ export class SDFGRenderer extends EventEmitter {
...
+
+        // Expand by one level
```
Comment (tbennun) [inline src/renderer/renderer.ts:876]: I'm not sure about this mode for simple SDFGs, where we will start with states expanded but contents of maps collapsed. This might be good for huge graphs but some kernels will just require users to double click nodes, which some users do not know they can. @phschaad Thoughts?

**Comment by LeandroTreu** [comment]
> I think that the overlay LOD management needs to be refactored out to some node overlay superclass.

Yes, Philipp and I already discussed this. At the moment with my thesis I focus on the performance optimizations, and I checked the Overlay system in the profiler and only found little overhead with the current setup. Depending on the time I have left after the main part of my thesis, refactoring the Overlays could be an additional task I could work on.

File: src/sdfv.ts (hunk 2 above)
Comment (LeandroTreu) [reply src/sdfv.ts:533]: This should have been 10 ms from the beginning like all the other setTimeout() I added. Does not make a difference as 10 ms is enough for the browser to schedule the function for later.

File: src/renderer/renderer.ts (hunk 3 above)
Comment (LeandroTreu) [reply src/renderer/renderer.ts:874]: Oh yes this was a quick fix because there was an error without the this.relayout() in-between the fully collapse and expand. I just took the existing collapse and expand code from the renderer. Relayouting after fully collapsing here only takes a millisecond, so if it is really necessary it's only a small overhead. I will check again though!

File: src/renderer/renderer.ts (hunk 4 above)
Comment (LeandroTreu) [reply src/renderer/renderer.ts:876]: I could expand graphs depending on how large they are or add "expand" symbols to collapsed nodes. The main reasons for starting fully collapsed are the loading time when opening a file (54 seconds for the 255 MiB graph on my laptop) and readability of deeply nested graphs.

**Comment by LeandroTreu** [comment]
Branch is NOT READY FOR MERGE YET as there is still a runtime error. Also I still need to look at the startup collapse & expand.

**Comment by LeandroTreu** [comment]
Need to first merge april fixes 2 to fix the runtime error, then this pull request is ready to merge into master. Collapsed startup is commented out for reworking it in a later commit.

File: src/renderer/renderer.ts [hunk 5]
```diff
@@ -901,10 +902,36 @@ export class SDFGRenderer extends EventEmitter {
         else
             this.bgcolor = window.getComputedStyle(this.canvas).backgroundColor;
 
-        // Create the initial SDFG layout
-        // Loading animation already started in the file_read_complete function
-        // in sdfv.ts to also include the JSON parsing step.
+
         this.updateCFGList();
+
+        // Create the initial SDFG layout
```
Comment (phschaad) [inline src/renderer/renderer.ts:908]: Since we're leaving this out for now, please remove this commented out section for this PR

File: src/renderer/renderer.ts [hunk 6]
```diff
@@ -3043,7 +3119,7 @@ export class SDFGRenderer extends EventEmitter {
                             highlighting_changed = true;
                             hover_changed = true;
                         }
-
+
```
Comment (phschaad) [inline src/renderer/renderer.ts:3122]: fmt

**Review by phschaad — CHANGES_REQUESTED** [review]
LGTM in general, only two small corrections. Also it would be really nice if the PR description had an image demoing the edge summary :-)

**Review by phschaad — APPROVED** [review]
